elements/lisk-dpos/src/random_seed.ts

```diff
@@ -0,0 +1,221 @@
+
+	// tslint:disable-next-line:no-let
+	for (let i = 0; i < outputSize; i += 1) {
+		// tslint:disable-next-line:no-bitwise
+		result[i] = bufferArray.map(b => b[i]).reduce((a, b) => a ^ b, 0);
+	}
+
+	return result;
+};
+
+const findPreviousHeaderOfDelegate = (
+	header: BlockHeader,
+	searchTillHeight: number,
+	headersMap: HeadersMap,
+): BlockHeader | undefined => {
+	const { height, generatorPublicKey } = header;
+	const searchTill = Math.max(searchTillHeight, 1);
+
+	// tslint:disable-next-line:no-let
+	for (let i = height - 1; i >= searchTill; i -= 1) {
+		if (headersMap[i].generatorPublicKey === generatorPublicKey) {
+			return headersMap[i];
+		}
+	}
+
+	return undefined;
+};
+
+const isValidSeedReveal = (
+	seedReveal: string,
+	previousSeedReveal: string,
+): boolean =>
+	bufferToHex(strippedHash(hexToBuffer(seedReveal))) === previousSeedReveal;
+
+const selectSeedReveals = ({
+	fromHeight,
+	toHeight,
+	headersMap,
+	rounds,
+}: {
+	readonly fromHeight: number;
+	readonly toHeight: number;
+	readonly headersMap: HeadersMap;
+	readonly rounds: Rounds;
+}): Buffer[] => {
+	const selected = [];
+
+	// tslint:disable-next-line:no-let
+	for (let i = fromHeight; i >= toHeight; i -= 1) {
+		const header = headersMap[i];
+		const blockRound = rounds.calcRound(header.height);
+
+		const lastForgedBlock = findPreviousHeaderOfDelegate(
+			header,
+			rounds.calcRoundStartHeight(blockRound - 1),
+			headersMap,
+		);
+
+		// If delegate not forged any other block earlier in current and last round
+		if (!lastForgedBlock) {
+			continue;
+		}
+
+		// To validate seed reveal of any block in the last round
+		// We have to check till second last round
+		if (!isValidSeedReveal(header.seedReveal, lastForgedBlock.seedReveal)) {
+			continue;
+		}
+
+		selected.push(hexToBuffer(header.seedReveal));
+	}
+
+	return selected;
+};
+
+export const generateRandomSeeds = (
+	round: number,
+	rounds: Rounds,
+	headers: ReadonlyArray<BlockHeader>,
+	// tslint:disable-next-line:no-magic-numbers
+): FixedLengthArray<RandomSeed, 2> => {
+	// Middle range of a round to validate
+	// tslint:disable-next-line:no-magic-numbers
+	const middleThreshold = Math.floor(rounds.blocksPerRound / 2);
+	const currentRound = round;
+	const lastBlockHeight = headers[headers.length - 1].height;
+	const startOfCurrentRound = rounds.calcRoundStartHeight(currentRound);
+	const middleOfCurrentRound = rounds.calcRoundMiddleHeight(currentRound);
+	const startOfLastRound = rounds.calcRoundStartHeight(currentRound - 1);
+	const endOfLastRound = rounds.calcRoundEndHeight(currentRound - 1);
+	// tslint:disable-next-line:no-magic-numbers
+	const startOfSecondLastRound = rounds.calcRoundStartHeight(currentRound - 2);
+
+	if (lastBlockHeight < middleOfCurrentRound) {
+		throw new Error(
+			`Random seed can't be calculated earlier in a round. Wait till you pass middle of round. Current height: ${lastBlockHeight}`,
+		);
+	}
+
+	if (currentRound === 1) {
```

Comment:
I think this part is already done in the genesis block processing, so it will never come here?

Reply:
Yes based on `forger_info.ts` usage it will not reach here. But based on the LIP specification, one can generate random seed after block height 51, even in the first round and will not be zero seed.